lisa/sut_orchestrator/baremetal/source.py

```diff
@@ -0,0 +1,81 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import re
+import zipfile
+from typing import Type
+
+import requests
+from azure.devops.connection import Connection  # type: ignore
+from msrest.authentication import BasicAuthentication
+
+from lisa import schema
+from lisa.util import InitializableMixin, constants, get_matched_str, subclasses
+from lisa.util.logger import get_logger
+
+from .schema import ADOSourceSchema, SourceSchema
+
+
+class Source(subclasses.BaseClassWithRunbookMixin, InitializableMixin):
+    def __init__(self, runbook: SourceSchema) -> None:
+        super().__init__(runbook=runbook)
+        self._log = get_logger("source", self.__class__.__name__)
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return SourceSchema
+
+    def download(self) -> str:
+        raise NotImplementedError()
+
+
+class ADOSource(Source):
+    __file_format = re.compile(r"format=(?P<format>.*)", re.M)
+
+    def __init__(self, runbook: ADOSourceSchema) -> None:
+        super().__init__(runbook)
+        self.ado_runbook: ADOSourceSchema = self.runbook
+        self._log = get_logger("ado", self.__class__.__name__)
+
+    @classmethod
+    def type_name(cls) -> str:
+        return "ado"
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return ADOSourceSchema
+
+    def download(self) -> str:
```

Comment:
The different source may need different implementation, and to be called in build class.